packages/bridge-ui-v2/src/libs/proof/BridgeProver.ts

```diff
@@ -1,27 +1,213 @@
+    }
+
+    // RLP encode the proof together for LibTrieProof to decode
+    const rlpEncodedStorageProof = toRlp(proof.storageProof[0].proof);
+
+    return { proof, rlpEncodedStorageProof };
   }
 
   // Reference: EncodedSignalProof in relayer/proof/encoded_signal_proof.go
   // protocol/contracts/signal/SignalService.sol
-  private _encodedSignalProof(crossChainSyncAddress: Address, rlpEncodedStorageProof: Hex, blockHeight: bigint) {
-    type Hop = {
-      signalRootRelay: Address;
-      signalRoot: Hex;
-      storageProof: Hex;
-    };
-
-    const hops: Hop[] = [];
-    const signalProof = encodeAbiParameters(
+  async encodedSignalProof(msgHash: Hash, srcChainId: number, destChainId: number) {
+    const srcBridgeAddress = routingContractsMap[srcChainId][destChainId].bridgeAddress;
+    const srcSignalServiceAddress = routingContractsMap[srcChainId][destChainId].signalServiceAddress;
+    const destCrossChainSyncAddress = routingContractsMap[destChainId][srcChainId].taikoAddress;
+
+    // Get the block from chain A based on the latest block hash
+    // we get cross chain (Taiko contract on chain B)
+    const blockNumber = await this.getBlockNumber(srcChainId, destChainId, destCrossChainSyncAddress);
+
+    const { rlpEncodedStorageProof } = await this.encodedStorageProof({
+      msgHash,
+      clientChainId: srcChainId,
+      contractAddress: srcBridgeAddress,
+      proofForAccountAddress: srcSignalServiceAddress,
+      blockNumber,
+    });
+
+    const signalProof = this._encodeAbiParameters(
+      destCrossChainSyncAddress,
+      BigInt(blockNumber),
+      rlpEncodedStorageProof,
+      [],
+    );
+    return signalProof;
+  }
+
+  async encodedSignalProofWithHops(
+    msgHash: Hash,
+    receipt: TransactionReceipt,
+    srcChainId: number,
+    destChainId: number,
+  ) {
+    const srcBridgeAddress = routingContractsMap[srcChainId][destChainId].bridgeAddress;
+    const srcSignalServiceAddress = routingContractsMap[srcChainId][destChainId].signalServiceAddress;
+    const destCrossChainSyncAddress = routingContractsMap[destChainId][srcChainId].taikoAddress;
+    const hopParams = routingContractsMap[srcChainId][destChainId].hops;
+
+    let blockNumber: number = 0;
+    // Initialize hopChainId with src chain
+    let hopChainId: number = srcChainId;
+    for (var hop of hopParams) {
+      blockNumber = await this.getBlockNumber(hopChainId, hop.chainId, hop.taikoAddress);
+      hopChainId = hop.chainId;
+    }
+    // Get the block number from last hop chain to dest chain
+    blockNumber = await this.getBlockNumber(hopChainId, destChainId, destCrossChainSyncAddress);
+
+    // Generate main storage proof with receipt.blockNumber
+    const { proof, rlpEncodedStorageProof } = await this.encodedStorageProof({
+      msgHash,
+      clientChainId: srcChainId,
+      contractAddress: srcBridgeAddress,
+      proofForAccountAddress: srcSignalServiceAddress,
+      blockNumber: receipt.blockNumber,
```

Comment:
I see this the only place we actually need the receipt. From what I understood about the proofs, we don't necessarily need the exact block, just any after. Technically 'latest' works, but it could not be synced. So if we fetch the latest synced block, so the one you fetch above, doesn't this work as well?

Reply:
Good question. Let me verify it.

Reply:
We don't need the receipt here. Removed.